DefaultDnsServerAddressStreamProvider fails when /etc/resolv.conf is not present on ZOS

### Expected behavior
`DefaultDnsServerAddressStreamProvider` should fallback to the existing lookup if `/etc/resolv.conf` is not present

### Actual behavior
`IllegalStateException` that finally leads to a `NoClassDefFoundError` when `DnsServerAddresses.defaultAddressList()` is called on ZOS.

### Steps to reproduce
rename `/etc/resolv.conf` and call e.g. `DnsServerAddresses.defaultAddressList()`

### Minimal yet complete reproducer code (or URL to code)

### Netty version
Issue was introduced with https://github.com/netty/netty/pull/13884, so `4.1.108.Final`

### JVM version (e.g. `java -version`)
```
java version "1.8.0_391"
Java(TM) SE Runtime Environment (build 8.0.8.15 - pmz3180sr8fp15-20231030_01(SR8 FP15))
IBM J9 VM (build 2.9, JRE 1.8.0 z/OS s390-31-Bit 20230908_57744 (JIT enabled, AOT enabled)
OpenJ9  - ec90c19
OMR   - 8b64d40
IBM   - 696e9df)
JCL - 20231012_01 based on Oracle jdk8u391-b13
```

### OS version (e.g. `uname -a`)
```
OS/390 TVT6016 28.00 04 2964
```

Here is the log on our System that shows the `IllegalStateException` and the `NoClassDefFoundError`

```
2024-04-04T01:38:56.515-04:00 | ERROR | 341-4c29-4836-a486-1617769ae446) | Host             | com.instana.sensor-host - 1.1.171 | Error on activate
java.lang.NoClassDefFoundError: io.netty.resolver.dns.DefaultDnsServerAddressStreamProvider (initialization failure)
	at java.lang.J9VMInternals.initializationAlreadyFailed(J9VMInternals.java:107) ~[?:2.9 (09-08-2023)]
	at io.netty.resolver.dns.DnsServerAddresses.defaultAddressList(DnsServerAddresses.java:41) ~[?:?]
	at com.instana.agent.host.sensor.impl.reversedns.ReverseIPv4Lookup.getInstance(ReverseIPv4Lookup.java:15) ~[?:?]
	at com.instana.agent.host.sensor.Host.activate(Host.java:173) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90) ~[?:1.8.0]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55) ~[?:1.8.0]
	at java.lang.reflect.Method.invoke(Method.java:508) ~[?:1.8.0]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:244) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:685) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:529) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:318) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:308) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:354) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:1000) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:973) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:776) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:674) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:437) ~[?:?]
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.configurationUpdated(ConfigurableComponentHolder.java:442) ~[?:?]
	at org.apache.felix.scr.impl.manager.RegionConfigurationSupport.configurationEvent(RegionConfigurationSupport.java:347) ~[?:?]
	at org.apache.felix.scr.impl.manager.RegionConfigurationSupport$2.configurationEvent(RegionConfigurationSupport.java:115) ~[?:?]
	at org.apache.felix.cm.impl.ConfigurationManager$FireConfigurationEvent.sendEvent(ConfigurationManager.java:1721) ~[?:?]
	at org.apache.felix.cm.impl.ConfigurationManager$FireConfigurationEvent.run(ConfigurationManager.java:1662) ~[?:?]
	at org.apache.felix.cm.impl.UpdateThread.run0(UpdateThread.java:122) ~[?:?]
	at org.apache.felix.cm.impl.UpdateThread.run(UpdateThread.java:84) ~[?:?]
	at java.lang.Thread.run(Thread.java:825) [?:2.9 (09-08-2023)]
Caused by: java.lang.IllegalStateException: /etc/resolv.conf could not be read
	at io.netty.resolver.dns.ResolvConf.system(ResolvConf.java:66) ~[?:?]
	at io.netty.resolver.dns.DefaultDnsServerAddressStreamProvider.<clinit>(DefaultDnsServerAddressStreamProvider.java:57) ~[?:?]
	at io.netty.resolver.dns.DnsServerAddresses.defaultAddressList(DnsServerAddresses.java:41) ~[?:?]
	at com.instana.agent.host.sensor.impl.reversedns.ReverseIPv4Lookup.getInstance(ReverseIPv4Lookup.java:15) ~[?:?]
	at com.instana.agent.host.sensor.Host.activate(Host.java:173) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90) ~[?:1.8.0]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55) ~[?:1.8.0]
	at java.lang.reflect.Method.invoke(Method.java:508) ~[?:1.8.0]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:244) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:685) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:529) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:318) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:308) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:354) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:1000) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:973) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:918) ~[?:?]
	at org.apache.felix.framework.ServiceRegistrationImpl.getFactoryUnchecked(ServiceRegistrationImpl.java:347) ~[?:?]
	at org.apache.felix.framework.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:247) ~[?:?]
	at org.apache.felix.framework.ServiceRegistry.getService(ServiceRegistry.java:350) ~[?:?]
	at org.apache.felix.framework.Felix.getService(Felix.java:3737) ~[?:?]
	at org.apache.felix.framework.BundleContextImpl.getService(BundleContextImpl.java:470) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleRefPair.getServiceObject(SingleRefPair.java:88) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BindMethod.getServiceObject(BindMethod.java:675) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager.getServiceObject(DependencyManager.java:2556) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager.doInvokeBindMethod(DependencyManager.java:2075) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager.invokeBindMethod(DependencyManager.java:2058) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.invokeBindMethod(SingleComponentManager.java:443) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:333) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:301) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1200) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1121) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.trackAdding(ServiceTracker.java:928) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.track(ServiceTracker.java:864) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:1152) ~[?:?]
	at org.apache.felix.scr.impl.BundleComponentActivator$ListenerInfo.serviceChanged(BundleComponentActivator.java:114) ~[?:?]
	at org.apache.felix.framework.EventDispatcher.invokeServiceListenerCallback(EventDispatcher.java:990) ~[?:?]
	at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:838) ~[?:?]
	at org.apache.felix.framework.EventDispatcher.fireServiceEvent(EventDispatcher.java:545) ~[?:?]
	at org.apache.felix.framework.Felix.fireServiceEvent(Felix.java:4595) ~[?:?]
	at org.apache.felix.framework.Felix.registerService(Felix.java:3587) ~[?:?]
	at org.apache.felix.framework.BundleContextImpl.registerService(BundleContextImpl.java:348) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:929) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:915) ~[?:?]
	at org.apache.felix.scr.impl.manager.RegistrationManager.changeRegistration(RegistrationManager.java:133) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.registerService(AbstractComponentManager.java:984) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:752) ~[?:?]
	... 10 more
```

--- a/resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsServerAddressStreamProvider.java
+++ b/resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsServerAddressStreamProvider.java
@@ -50,11 +50,22 @@ public final class DefaultDnsServerAddressStreamProvider implements DnsServerAdd
     static {
         final List<InetSocketAddress> defaultNameServers = new ArrayList<InetSocketAddress>(2);
         if (!PlatformDependent.isAndroid()) {
-            // Only try to use when not on Android as the classes not exists there:
+            // Skip this on Android; it has neither /etc/resolv.conf nor JNDI classes.
             // See https://github.com/netty/netty/issues/8654
             if (!PlatformDependent.isWindows()) {
-                // /etc/resolv.conf exists on Linux + macOS, but not on Android
-                defaultNameServers.addAll(ResolvConf.system().getNameservers());
+                // Try reading /etc/resolv.conf. It's usually found on Linux or macOS, but can also be missing.
+                try {
+                    defaultNameServers.addAll(ResolvConf.system().getNameservers());
+                } catch (IllegalStateException e) {
+                    String fallbackMessage = "Failed to get name servers from /etc/resolv.conf; will fall back to JNDI";
+                    if (logger.isDebugEnabled()) {
+                        // Always log at INFO, but only include stack trace if DEBUG is enabled.
+                        logger.info(fallbackMessage, e);
+                    } else {
+                        logger.info(fallbackMessage);
+                    }
+                    DirContextUtils.addNameServers(defaultNameServers, DNS_PORT);
+                }
             } else {
                 DirContextUtils.addNameServers(defaultNameServers, DNS_PORT);
             }

